src/typechecker.erl

```diff
@@ -3408,14 +3408,54 @@ refinable(_) ->
 no_guards({clause, _, _, Guards, _}) ->
     Guards == [].
 
+
+-spec check_guard_bif(erlang:anno(), atom(), list()) -> map().
+check_guard_bif(P, is_atom, [{var, _, Var}]) -> #{Var => {type, P, atom, []}};
+check_guard_bif(P, is_binary, [{var, _, Var}]) -> #{Var => {type, P, binary, []}};
+check_guard_bif(P, is_bitstring, [{var, _, Var}]) -> #{Var => {type, P, bitstring, []}};
+check_guard_bif(P, is_boolean, [{var, _, Var}]) -> #{Var => {type, P, boolean, []}};
+check_guard_bif(P, is_float, [{var, _, Var}]) -> #{Var => {type, P, float, []}};
+check_guard_bif(P, is_integer, [{var, _, Var}]) -> #{Var => {type, P, integer, []}};
+check_guard_bif(P, is_number, [{var, _, Var}]) -> #{Var => {type, P, number, []}};
+check_guard_bif(P, is_list, [{var, _, Var}]) -> #{Var => {type, P, list, []}};
+check_guard_bif(P, is_map, [{var, _, Var}]) -> #{Var => {type, P, map, []}};
+check_guard_bif(P, is_pid, [{var, _, Var}]) -> #{Var => {type, P, pid, []}};
+check_guard_bif(P, is_port, [{var, _, Var}]) -> #{Var => {type, P, port, []}};
+check_guard_bif(P, is_reference, [{var, _, Var}]) -> #{Var => {type, P, reference, []}};
+check_guard_bif(P, is_tuple, [{var, _, Var}]) -> #{Var => {type, P, tuple, []}};
+check_guard_bif(P, is_function, [{var, _, Var}]) -> #{Var => {type, P, 'fun', []}};
+check_guard_bif(P, is_record, [{var, _, Var}, {atom, _, Record}]) ->
+    #{Var => {type, P, record, [{atom, P, Record}]}};
+check_guard_bif(_P, _Fun, _Vars) ->
+    #{}.
+
+-spec check_guard_call(erlang:anno(), atom(), list()) -> map().
+check_guard_call(P, Name, Args) ->
+    Arity = length(Args),
+    case erl_internal:bif(Name, Arity) of
+        true -> check_guard_bif(P, Name, Args);
+        false -> #{}
+    end.
+
+-spec check_guard(#env{}, term()) -> map().
+check_guard(_Env, {call, P, {atom, _, Fun}, Vars}) ->
+    check_guard_call(P, Fun, Vars);
+check_guard(_Env, {call, P, {remote,_,_,{atom, _, Fun}}, Vars}) ->
+    check_guard_call(P, Fun, Vars);
+check_guard(Env, {op, _OrElseAnno, 'orelse', Call1, Call2}) ->
+    union_var_binds([check_guard(Env, Call1), check_guard(Env, Call2)], Env#env.tenv);
+check_guard(Env, Guard) ->
+    {_Ty, VB, _Cs} = type_check_expr(Env, Guard), % Do we need to thread the Env?
+    VB.
+
 %% TODO: implement proper checking of guards.
+-spec check_guards(#env{}, term()) -> map().
 check_guards(Env, Guards) ->
     union_var_binds(
       lists:map(fun (GuardSeq) ->
                         union_var_binds(
                           lists:map(fun (Guard) ->
-                                                {_Ty, VB, _Cs} = type_check_expr(Env, Guard), % Do we need to thread the Env?
-                                                VB
+                              check_guard(Env, Guard)
                                     end, GuardSeq), Env#env.tenv)
                 end, Guards), Env#env.tenv).
```

Comment:
There the nested semicolon-separated list of comma-separated guards is traversed. I can't see that the or/and logic for guards is handled here. I'm just worried about OR-guard sequences like `when is_integer(X); is_atom(X)` and `when is_integer(X); is_integer(Y)`. In the first case, it's possible to refine X to `integer() | atom()` but in the second case we can't refine X nor Y. In cases where we can't be sure, its better to not refine at all IMO, so perhaps you should limit the refinement to guard sequences without semicolons (i.e. singleton list on the outer level) for now?

@NelsonVides made #206 and it had the same issue initially. Don't know if it was fixed later.

Reply:
I think this is what I tried implementing here: https://github.com/josefs/Gradualizer/blob/691a3bc4d15ef877a16b6c546290c22d398a2199/src/typechecker.erl#L3431-L3450

Reply:
Yea thanks for that. Used some of your logic in my update.

Reply:
Can we resolve this? Handled both `orelse` and `andalso` in tests.

Reply:
Yeap, good.